edgelet/edgelet-http/tests/tls.rs

```diff
@@ -0,0 +1,112 @@
+
+use edgelet_core::crypto::CreateCertificate;
+use edgelet_core::{CertificateIssuer, CertificateProperties, CertificateType, IOTEDGED_CA_ALIAS};
+use edgelet_hsm::Crypto;
+use edgelet_http::certificate_manager::CertificateManager;
+use edgelet_http::route::{Builder, Parameters, RegexRoutesBuilder, Router};
+use edgelet_http::Error as HttpError;
+use edgelet_http::HyperExt;
+use edgelet_http::{Run, Version};
+
+use futures::{future, Future};
+use hyper::server::conn::Http;
+use hyper::{Body, Request, Response, StatusCode};
+use hyper_tls;
+use native_tls::TlsConnector;
+use tempdir::TempDir;
+use url::Url;
+
+const HOMEDIR_KEY: &str = "IOTEDGE_HOMEDIR";
+
+#[test]
+fn tls_functional_test() {
+    let mut http = hyper::client::HttpConnector::new(1);
+    http.enforce_http(false);
+
+    let mut tls_connector_builder = TlsConnector::builder();
+
+    // This is because are using a self signed cert
+    tls_connector_builder.danger_accept_invalid_certs(true);
+
+    let tls_connector = tls_connector_builder.build().unwrap();
+
+    let https_connector =
+        hyper_tls::HttpsConnector::<hyper::client::HttpConnector>::from((http, tls_connector));
+
+    let client = hyper::Client::builder().build::<_, hyper::Body>(https_connector);
+
+    let request = client.get(
+        "https://localhost:8881/route1/hello?api-version=2018-06-28"
+            .parse()
+            .unwrap(),
+    );
+
+    let mut runtime = tokio::runtime::current_thread::Runtime::new().unwrap();
+
+    let server = configure_test().map_err(|err| eprintln!("{}", err));
+    runtime.spawn(server);
+
+    let res = runtime.block_on(request).unwrap();
+    assert_eq!(res.status(), 200);
+}
+
+pub fn configure_test() -> Run {
+    // setup the IOTEDGE_HOMEDIR folder where certs can be generated and stored
+    let home_dir = TempDir::new("tls_integration_test").unwrap();
+    env::set_var(HOMEDIR_KEY, &home_dir.path());
+    println!("IOTEDGE_HOMEDIR set to {:#?}", home_dir.path());
+
+    let crypto = Crypto::new().unwrap();
+
+    // create the default issuing CA cert properties
+    let edgelet_ca_props = CertificateProperties::new(
+        3600,
+        "test-iotedge-cn".to_string(),
+        CertificateType::Ca,
+        IOTEDGED_CA_ALIAS.to_string(),
+    )
+    .with_issuer(CertificateIssuer::DeviceCa);
+
+    let _workload_ca_cert = crypto.create_certificate(&edgelet_ca_props).unwrap();
+
+    let edgelet_cert_props = CertificateProperties::new(
+        3600,
+        "testtls".to_string(),
+        CertificateType::Server,
+        "localhost".to_string(),
+    )
+    .with_issuer(CertificateIssuer::DeviceCa);
+
+    let manager = CertificateManager::new(crypto.clone(), edgelet_cert_props);
+
+    let recognizer = RegexRoutesBuilder::default()
+        .get(Version::Version2018_06_28, "/route1/hello", route1)
+        .finish();
+    let router = Router::from(recognizer);
+
+    Http::new()
+        .bind_url(
+            Url::parse("https://127.0.0.1:8881").unwrap(),
```

Comment:
Should we use an arbitrary unused TCP port instead of `8881`? See `edgelet_test_utils::get_unused_tcp_port`.